### ISSUE TITLE
Is it a bug that jl_set_const C api allows changing a non-const symbol into a const one?

### ISSUE THREAD

**Issue by NHDaly** [JuliaLang/julia#28106]
The current code for `jl_set_const` will change a non-const global variable into a const one:

https://github.com/JuliaLang/julia/blob/a5a4e9901d6c4462f685ac2903c2a39d32c50b05/src/module.c#L491-L499

Julia doesn't let you do that in-language:
```julia
julia> x = 5
>> 5

julia> const x = 5
ERROR: cannot declare x constant; it already has a value
```

My understanding was that this is because a function could have been compiled expecting to have to dynamically check `x`'s type, but if it's a `const`, it doesn't store that information. Is that right? And if so, is this C api behavior a bug?

(I've tested it, and `jl_set_const` will happily change a non-const to const, as described.)

**Comment by JeffBezanson** [comment]
Yes, `bp->value == NULL` should be added as a precondition of that function, in whatever way we decide to handle that.

**Comment by NHDaly** [comment]
Cool, thanks.

Actually, this could maybe just delegate to `jl_declare_constant`, which seems to already correctly check for this. Maybe I can redefine it as:
```C
 JL_DLLEXPORT void jl_set_const(jl_module_t *m, jl_sym_t *var, jl_value_t *val) 
 { 
     jl_binding_t *bp = jl_get_binding_wr(m, var, 1); 
     jl_declare_constant(bp);
     bp->value = val; 
     jl_gc_wb(m, val); 
 } 
```

But idk, maybe it's better just to be explicit, even if there's some code duplication. Let me know what you think and i can send a PR! :)

**Comment by adienes** [comment]
now has an extremely scary warning

https://github.com/JuliaLang/julia/blob/8ddf7d8f73eb3bc1f53bda70d55f318e7ed6f92e/src/module.c#L1767-L1770

and a safer alternative in`jl_declare_constant_val`
